src/3D/Camera.cpp

```diff
@@ -9,28 +9,97 @@
 
 #include "Camera.h"
 #include "3D/LandIsland.h"
+#include "ECS/Components/Transform.h"
 #include "ECS/Registry.h"
 #include "Game.h"
 
 #include <glm/gtx/euler_angles.hpp>
+#include <glm/gtx/intersect.hpp>
+#include <glm/gtx/norm.hpp>
+
+#define OUT
 
 using namespace openblack;
 
-glm::mat4 Camera::getRotationMatrix() const
+Camera::Camera(glm::vec3 position, glm::vec3 rotation)
+    : _position(position)
+    , _rotation(glm::radians(rotation))
+    , _dv(0.0f, 0.0f, 0.0f)
+    , _dwv(0.0f, 0.0f, 0.0f)
+    , _dsv(0.0f, 0.0f, 0.0f)
+    , _ddv(0.0f, 0.0f, 0.0f)
+    , _duv(0.0f, 0.0f, 0.0f)
+    , _drv(0.0f, 0.0f, 0.0f)
+    , _projectionMatrix(1.0f)
+    , _velocity(0.0f, 0.0f, 0.0f)
+    , _hVelocity(0.0f, 0.0f, 0.0f)
+    , _rotVelocity(0.0f, 0.0f, 0.0f)
+    , _accelFactor(0.001f)
+    , _movementSpeed(4.0f)
+    , _maxMovementSpeed(0.005f)
+    , _maxRotationSpeed(0.005f)
+    , _mouseIsDown(false)
+    , _mouseIsMoving(false)
+    , _shiftHeld(false)
+{
+}
+
+glm::mat4 Camera::GetRotationMatrix() const
 {
 	return glm::eulerAngleZXY(_rotation.z, _rotation.x, _rotation.y);
 }
 
 glm::mat4 Camera::GetViewMatrix() const
 {
-	return getRotationMatrix() * glm::translate(glm::mat4(1.0f), -_position);
+	return GetRotationMatrix() * glm::translate(glm::mat4(1.0f), -_position);
 }
 
 glm::mat4 Camera::GetViewProjectionMatrix() const
 {
 	return GetProjectionMatrix() * GetViewMatrix();
 }
 
+bool Camera::RaycastCamToLand(OUT glm::vec3* hit)
+{
+	// get the hit by raycasting to the land down the camera ray
+	float intersectDistance = 0.0f;
+	glm::vec3 viewVec = GetForward();
+	if (glm::intersectRayPlane(_position, viewVec, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f), intersectDistance))
+	{
+		*hit = _position + viewVec * intersectDistance;
+		hit->y = Game::instance()->GetLandIsland().GetHeightAt(glm::vec2(hit->x, hit->z));
+		return true;
+	}
+	else
+	{
+		*hit = _position + viewVec * 100.f;
+	}
+	return false;
+}
+
+bool Camera::RaycastMouseToLand(OUT glm::vec3* hit)
+{
+	// get the hit by raycasting to the land down via the mouse
+	float intersectDistance = 0.0f;
+	int sWidth, sHeight;
```

Comment:
These are not initialised but used below with the vec4. Causing an exception on my end.

Reply:
Warnings are also picking it up as potential bugs

Reply:
this was a good catch, they are not initialized at all! I had forgotten a line!